### PR TITLE
[Example] Add Top-n-sigma custom logit processor

### LINKED ISSUE #29397: Add Top-nσ logit truncation custom logit processor example
## Motivation

SGLang's custom logit processor framework (enabled via `--enable-custom-logit-processor`) lets users inject arbitrary logit-space processing per-request. However, the only existing reference is a **dummy** `DeterministicLogitProcessor` in the docs that always samples a fixed token — there is no example of a processor that does meaningful logit filtering.

This issue proposes adding the **first real-world custom logit processor example**: [Top-nσ](https://arxiv.org/abs/2411.07641) ("Top-nσ: Not All Logits Are You Need", ACL 2025), a logit-space dynamic truncation method.

```
threshold = max_logit - n * std_logit
logits[logits < threshold] = -inf
```

Unlike sglang's built-in probability-space filters (`top_k`, `top_p`, `min_p`), Top-nσ operates **before softmax** (where SGLang's custom processor pipeline runs) and adapts the candidate set to the "peakiness" of the distribution.

### How it compares to existing methods

| Method | Space | Truncation rule | Adaptive? |
|---|---|---|---|
| top_k | Probability | Keep top-k highest-probability tokens | ❌ Fixed count |
| top_p (nucleus) | Probability | Keep tokens until cumulative prob ≥ p | ✅ Adaptive count only |
| min_p | Probability | Keep tokens with prob ≥ max_prob × min_p | ✅ Relative threshold |
| top_nσ | **Logit** | threshold = max_logit − n × std | ✅ Dual-adaptive |

> Key distinction: sglang's top_k/top_p/min_p all operate **after softmax** (probability space); top_nσ operates **before softmax** (logit space). Top-nσ is the first logit-space truncation reference for the custom processor framework.

### Why Top-nσ?

- **Truncation before softmax** — filters in logit space first, then softmax only processes the surviving candidates.
- **Dual adaptivity** — the std-based threshold adjusts both *how many* and *how loosely* candidates are kept in one formula.
- **Temperature-stable** — candidate count stays consistent across temperature scaling (logit std scales proportionally).
- **Argmax-invariant** — the top-1 token always survives, so greedy decoding is unaffected.

## Proposal

Add an example in a new `examples/runtime/custom_logit_processor/` folder — `top_n_sigma.py` — following the existing `examples/runtime/` pattern (standalone script, documented server launch command, client-side usage).

**No framework changes**: the processor is defined in the example file and injected per-request via the existing `custom_logit_processor` field + `--enable-custom-logit-processor` gate.

### Implementation characteristics

- **Fully vectorized** across the batch (no per-row Python loop)
- **std computed once** and reused for both the `std==0` guard and the threshold formula (avoids a duplicate vocab-wide reduction)
- **Argmax-invariant** by construction

### Edge cases handled

| Case | Behavior |
|------|----------|
| `top_n_sigma <= 0` | Request skipped |
| `top_n_sigma` is None / non-numeric | Request skipped |
| `std == 0` (all logits equal) | Row skipped (no filtering) |
| NaN/Inf logits | Row skipped (finite guard) |
| Empty / None param list | Logits returned unchanged |

## References

Tang et al., "Top-nσ: Not All Logits Are You Need", ACL 2025. [arXiv:2411.07641](https://arxiv.org/abs/2411.07641)

### PR DESCRIPTION
**Pull request by ajinkyajawale14499**
## Motivation

Closes #29397.

SGLang's custom logit processor framework (`--enable-custom-logit-processor`) currently only ships a **dummy** `DeterministicLogitProcessor` reference, which always samples a fixed token — there is no example of a processor that does meaningful logit filtering. This adds the first **real-world, logit-space** example: Top-n-sigma ("Top-nσ: Not All Logits Are You Need", Tang et al., ACL 2025, [arXiv:2411.07641](https://arxiv.org/abs/2411.07641)):

```
threshold = max_logit - n * std_logit
logits[logits < threshold] = -inf
```

Unlike the built-in probability-space filters (`top_k` / `top_p` / `min_p`) that run after softmax, Top-n-sigma runs *before* softmax — where the custom processor pipeline executes — and adapts the candidate set to the "peakiness" of the distribution.

## Modifications

- New file `examples/runtime/custom_logit_processor/top_n_sigma.py`. **No framework changes** — the processor is defined in the example and injected per-request via the existing `custom_logit_processor` field + `--enable-custom-logit-processor` gate.
- `TopNSigmaLogitProcessor(CustomLogitProcessor)`:
  - **Fully vectorized** across the batch (no per-row Python loop over logits).
  - **std computed once** and reused for both the `std == 0` guard and the threshold.
  - **Argmax-invariant** by construction — the max-logit token always clears the threshold, so greedy decoding is unaffected.
- Edge cases handled per the issue: `top_n_sigma <= 0` / `None` / non-numeric / `bool` → row skipped; `std == 0` → row skipped; non-finite logits → row skipped; empty / `None` param list → logits returned unchanged.
- Runnable demo: documented server launch command + a `/generate` request with and without the processor.

## Accuracy Tests

N/A to model forward — this is an example, not a change to kernels/forward code. The processor logic was verified locally against synthetic logits (torch): argmax-invariance, exact `threshold = max − n·std` masking, and the `std==0` / NaN-Inf / invalid-param / empty-list / per-row mixed-`n` cases — all pass.

## Speed Tests and Profiling

N/A — example addition; no inference-path code is modified.

## Checklist

- [x] Format your code according to [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit) — isort / ruff / black / codespell pass on the new file.
- [x] Add unit tests — N/A for an `examples/` script; logic verified locally as described above.
- [ ] Update documentation — N/A (happy to add a short note under the docs' "Custom logit processor" section if preferred).
- [ ] Accuracy and speed benchmarks — N/A (example).
- [x] Follow the SGLang code style guidance.

> Developed on macOS / Apple Silicon without the CUDA toolchain; the example targets a running server.

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #28275739123](https://github.com/sgl-project/sglang/actions/runs/28275739123)
Latest PR Test (Extra): :x: [Run #28275739046](https://github.com/sgl-project/sglang/actions/runs/28275739046)

### CODE REVIEW

**Comment by ajinkyajawale14499** [comment]
Implements #29397 — a vectorized Top-n-sigma custom logit processor example, no framework changes. pre-commit is green and the processor logic is verified locally (argmax-invariance, threshold math, and all the edge cases in the issue). As a first-time contributor I can't add the `run-ci` label myself — could a maintainer kindly run `/tag-and-rerun-ci`? Happy to adjust anything. cc @Jackie2049

**Comment by ajinkyajawale14499** [comment]
Closing in favor of #29398 (@Jackie2049's earlier PR for the same `top_n_sigma.py` example — it predates this one). I'll contribute the improvements (fp32 reductions, bool-guard, demo/test) over there. Thanks!
